rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,218 @@
+ignored. With the same backfill mechanism, the application should now be able to
+lower the retention height below the current base height (whether to accommodate
+changing consensus parameters, app specific needs, or a local config).
+This accommodates the use case of a truncated node that wishes to become an
+archive node.
+
+#### Choosing between Strong or Weak Gaurantees about Data
+
+It can be said previously, that block pruning, with respect to the application, 
+offered strong gaurantees about data i.e. if the application pruned to height 100, 
+the operation was synchronous and thus the application knew it would prune to 100 
+before the next height. This meant that the application always knew the height and 
+base of the blockchain at every step. The opposite, weak gaurantees, implies that 
+the application can't gaurantee that what it asked from Tendermint, with respect 
+to data, has happened.
+
+The downside of strong gaurantees is that it is blocking. If a validator decided 
+to prune a million blocks at once, the validator could potentially fall behind. 
+There's an argument then that pruning (and for that matter reverse sync) shouldn't 
+lie on the consensus critical path but rather be an asynchronous background process 
+of the node.
+
+Further to this argument, nodes will not be able to prune past the evidence age 
+regardless of the retain height that the application requests. Finally, applications 
+may ask to fetch blocks back to a height that none of the nodes peers have. 
+The mechanism must ensure termination even if it doesn't reach the height requested 
+by the app.
+
+Taking all these points into consideration, ReverseSync and pruning should be
+operations that run asynchronously in the background. Application can be informed
+of the current base via `RequestEndBlock`:
+
+```proto 
+message RequestEndBlock {
+  Height uint64
+  Base uint64
+}
+```
+
+### Verification
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. This then can be followed with
+using the hashes in the trusted header to validate the other block and state
+info. This is recursively done till the backfill height.  
+
+### Design
+
+This section will provide a high level picture of the design with the specifics being
+addressed in a following ADR. The reverse mechanism requires the gossiping of two new
+messages to allow the passsing of state data:
+
+```proto
+message StateDataRequest {
+  uint64 height = 1;
+}
+```
+
+```proto
+message StateDataResponse {
+  uint64 height = 1;
+  ValidatorSet validator_set = 2;
+  ConsensusParams consensus_params = 3;
+  ABCIResponses abci_responses = 4;
+}
+```
+
```

Comment:
Not convinced this should introduce new messages. We can reconstruct state from blocks, making this an implementation issue not protocol level issue. 

To me this seems like our implementation does not enable this, so we would like to change the protocol. The blockchain reactor can fetch blocks at any height, the difference being when backfilling, it should not replay the block on the app. I think if we were to refactor tendermint-go to enable cross pkg communication this problem would be solved. 

Just my 2¢

Reply:
> We can reconstruct state from blocks

Not going backwards I don't think - only forwards. We can't take state at height h-1 give it a block at height h and produce the corresponding state at height h

>  it should not replay the block on the app

How could you get `ABCI Responses` then if you didn't pass it to the application

Reply:
> Not going backwards I don't think - only forwards. We can't take state at height h-1 give it a block at height h and produce the corresponding state at height h

If you need height 100 you would need to fetch h + 1 as well, i believe 

> How could you get `ABCI Responses` then if you didn't pass it to the application

Oh when backfilling was mentioned to me before it was only for Tendermints case. I thought that is the approach we wanted.

Reply:
I guess the matter of ABCI responses is a separate concern. I too believe that in the long term we want to extract out ABCI Responses and Tx Indexer for that matter into its own third party thing. I would however leave it in for now and at a later date reassess its purpose.

Reply:
I would disagree. I am opposed to adding things now if we are not sure if it will be there in the future. This would mean two protocol breaking changes. I think going with the basic approach now (non-breaking) and asses in the future if users would like to have this data.

Reply:
> I think if we were to refactor tendermint-go to enable cross pkg communication this problem would be solved.

What do you mean by this?

Reply:
If consensus sees evidence it tells the blockchain reactor to fetch the needed data to verify it. I didnt want to use the term `reactor` communication because the communication could include non-reactors as well.

Reply:
> I am opposed to adding things now if we are not sure if it will be there in the future.

So the argument against this is a matter of completeness. It would be odd if a node had all the block data for a certain height yet didn't have the corresponding `ABCI Responses`.

As for the other state structures: `ValidatorSet` and `ConsensusParams`. We need this for evidence verification. There is no way currently that we can derive it from block data hence we need to introduce a new message to gossip it

Reply:
> So the argument against this is a matter of completeness. It would be odd if a node had all the block data for a certain height yet didn't have the corresponding ABCI Responses.

This sounds like the ABCI responses are being treated as part of a blocks state, just not in the state root. Are the ABCI responses merkelized into the header already, or do you have to re-execute the block to get them?

Reply:
Right, I looked into this a bit further and this is a valid point. I initially thought the `LastResultsHash` in the header could be used to verify `ABCIResponses`. But now I've just realised that the root hash is just for the `ResponseDeliverTxs`. 
```
LastResultsHash:                  ABCIResponsesResultsHash(abciResponses),
```
where
```
func ABCIResponsesResultsHash(ar *tmstate.ABCIResponses) []byte {
	return types.NewResults(ar.DeliverTxs).Hash()
}
```
Which means it doesn't include `BeginBlock` and `EndBlock` events. This is odd to me because I thought that it would have included all events that resulted from executing the block. Wouldn't this mean that if applications somehow fired different begin block and end block events that were not part of the `AppHash`, that Tendermint nodes would still agree on it? 🤔 

I would like to avoid executing prior blocks on the application. This, coupled with the growing sense that we shouldn't be saving `ABCIResponses` (as this is something that either the application or the TxIndexer should have/ be in control of), make me inclined to not worry about backfilling prior `ABCIResponses`. 

However, speaking of the txIndexer, I think it would be useful, when backfilling, to also batch the corresponding txs (`abci.TxResult`) to the TxIndexer. This means that although we won't need the entire `ABCIResponse`, we will at least need all the `DeliverTxResults`.